### PR TITLE
feat: Added Circular progress Widget

### LINKED ISSUE #10260: [Feature]: Circular progress widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Create a new circular progress widget

### Why should this be worked on?

[Community contributed](https://discord.com/channels/725602949748752515/827256420398399500/928270330768785469)

### PR DESCRIPTION
**Pull request by yaldram**
## Description
Added Widget for Circular Progress. Heavily Inspired by (copied from) - https://github.com/kevinsqi/react-circular-progressbar/tree/master/src

Fixes #10260

N / A

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- I have visually tested the widget from the app itself passing relevant props, please test on your side.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by yaldram** [comment]
- Please add an appropriate svg Icon.
- Also Please edit / change the helpTexts for the widget.

**Comment by shwetha-ramesh** [comment]
@somangshu Can we have the figma document for this widget.

**Comment by somangshu** [comment]
@shwetha-ramesh this widget is been contributed from community, We do not have designs here.

We however need design audit from @dilippitchika @momcilo-appsmith

**Comment by yaldram** [comment]
https://user-images.githubusercontent.com/13429535/148740600-c782bc32-6fff-470e-bf61-7f00e98af4f9.mp4

- Video recording demonstrating the functionality of the widget

**Comment by momcilo-appsmith** [comment]
A couple of questions that I have from testing this [Deploy Preview](https://appsmith-git-fork-yaldram-feature-circular-00cb58-get-appsmith.vercel.app):
1. Why is there a background color for the circle and why can't I remove it? (I'd suggest we remove it altogether)
2. Why are we defining success value?
3. Why is there a text color? (I'd suggest we remove it altogether)
4. Why is there background padding?

Suggestions on how to improve:
* For the default path color, I suggest: #EBEBEB
* I am pretty sure we should make this widget have the same property values as the progress bar we are working on.

I added the screenshot of its properties here. Except for the steps property, I am pretty sure we can reuse this and add the success color properties to it.
Screenshot 2022-01-10 at 12 15 03

**Comment by jsartisan** [comment]
Agreed with @momcilo-appsmith . The properties should be similar to what we have for the progress bar.

**Comment by yaldram** [comment]
Hey @momcilo-appsmith , my goal for this widget was to give the user all options for customization. For this widget I took inspiration from https://www.npmjs.com/package/react-circular-progressbar, and tried to incorporate a lot of cusomizations but we can for sure remove some properties if you feel they are not needed. To answer your questions : - 
1. Sure we can get rid of the backgroundColor altogether.
2. Well it was in part inspired by the Retool's completeColor property for their Circular Widget. So if you supply 100% their widget turns green and you can customize this completeColor. We have some more options, rather than a default 100% value you can customize the success value and can not only change the path color but also the text inside it. We can get rid of this as well if you want.
3. We can get rid of the Text color as well.
4. We can also get rid of the background padding.
5. For the path color we can have #EBEBEB as the default color.
6. I kind a disagree with this point, I think a circular progress widget is just a static widget I don't think we should show an intermediate form of it. On many dashboards and apps its job is just to show information like the percentage of sales this month, etc. So an intermediate value for progress means  you're not able to determine the value upfront.  I think we should just keep it simple and static.
7. Thanks for your feedback, I am developer so I like to build my components with a lot customizations but I totally agree with you we can remove a lot options from the pane and keep it simple & clean.

So we now have the following Options : - 
**Under General**
1. value
2. maxValue
3.  counterClockWise
4. strokeWidth

**Under Styles**
1. Path Color.
2. Track Color.

**Comment by yaldram** [comment]
Let me try implementing a Circular progress with Steps and I will get back to you. I will try matching it to the progress widget we already have.

**Comment by momcilo-appsmith** [comment]
@yaldram I was saying that we shouldn't implement steps to this circular bar actually.

**Comment by momcilo-appsmith** [comment]
@yaldram What I was trying to get into this widget (circular progress) is the determinate or indeterminate state where the first one has a value defined, and the second one doesn't (but is loading all the time until something is loaded).

**Comment by yaldram** [comment]
@momcilo-appsmith Ah Okay. should we do something like that ? For almost all the cases the widget would be a static one don't you think so ?

**Comment by yaldram** [comment]
Can you please do me a favour, can you please give me a list of all the configurations we must have in the pane so that I can remove others and update the widget. Thanks

**Comment by yaldram** [comment]
@momcilo-appsmith  I got what you mean say to show a loading like progress circle and the final bar when the value loads. Well I need to see if this is doable because it kind of delves into the state of the widget like say if a query is executing the widget is in intermediate state and if it returns a value it is in determinate state I don't know if it is doable in appsmith @jsartisan  please share your thoughts.

**Comment by jsartisan** [comment]
> For almost all the cases the widget would be a static one don't you think so ?

Actually no, the value can come from api or db query.

> Well I need to see if this is doable because it kind of delves into the state of the widget like say if a query is executing 
the widget is in intermediate state and if it returns a value it is in determinate state I don't know if it is doable in appsmith

That's actually very easy mate. Every appsmith widget gets a `isLoading` prop by default which is true when you bind something with api or query like `{{MyApi1.data}}` and that api/query is still loading.

**Comment by yaldram** [comment]
@jsartisan thanks for your response. I will try to come up with the loading state.

**Comment by shwetha-ramesh** [comment]
Moving back to Inprogress because of design changes.

**Comment by yaldram** [comment]
@jsartisan also I have a couple of queries for you : - 
1. Can you please leave a comment with all the props / config options that we need in the pane so that I can remove the others.
2. Also are we implementing the steps, determinate & indeterminate functionality like - 
image

**Comment by jsartisan** [comment]
> Can you please leave a comment with all the props / config options that we need in the pane so that I can remove the others.
We are currently discussing this. Will be back with a final answer in some moment.

> Also, are we implementing the steps, determinate & indeterminate functionality like 
In my opinion, we can skip version 1 of this widget.

**Comment by jsartisan** [comment]
@yaldram Here are a few changes that we will need before merging:

**Remove (or rather Hide) these - we will need defaults for these**
- Background, success, success text, and trail color.
- BG padding and text size - set to 0 and 20
- Max value, success value, and stroke width,

**Add**:

- Switch to show progress number. (Show result)
- Visible switch
- Change
- Change "value" to "Progress" (label change)
- Change "Path color" to "Fill color" (label change)

We need not add indeterminate right now.

cc: @dilippitchika

**Comment by somangshu** [comment]
Hey @yaldram, thanks for all the interest you have shown in widget development for Appsmith, It means a lot to us ❤️ And we love you for that 🤗 already 😄 

Please let us know in case you need any help from us on the last note given by @jsartisan, We would love it if this PR gets merged into our actual community edition. I am sure you'd feel proud too 🙂 💯

**Comment by dilippitchika** [comment]
@yaldram Adding the default values here for items we want you to hide

1. Background color - Transparent
2. Trail color - #E8E8E8
3. Success - #03B365
4. Success text - #231F20
5. BG Padding - 0
6. Text size - 20
7. Max value - 100
8. Success value - 100
9. Stroke width - 8

Thanks a lot for this, we are all still in awe of what you have built for us 🚀

**Comment by yaldram** [comment]
@somangshu  Thank you for the kind words means a a lot. @jsartisan thanks for your review. @dilippitchika  thanks for the values. Will refactor the code and push the changes in 2 - 3 days.

**Comment by yaldram** [comment]
Hello @jsartisan and @somangshu  I have pushed the changes please take a look. I also added the Separators / Steps please check let me know if you want to keep it or release it later. Thanks.

**Comment by yaldram** [comment]
https://user-images.githubusercontent.com/13429535/149463631-4392ad7a-3fb3-42fc-bf0e-ab7fa0f59707.mp4

**Comment by jsartisan** [comment]
@yaldram This looks good but there is one issue 😅 Our interminate/determinate feature is wrong in the progress bar widget and you replicated the same thing over here. Really sorry for that. Can you make the following changes:

1. Remove the `determinate/indeterminate` dropdown control for now. For now, we will just progress the bar without steps.
2. Since you will be removing the `determinate/indeterminate`, remove the `Steps` control too.
3. Remove the `Stroke Line Cap` because we will be managing the border radius with theming in future. For now, we can use the sharp border (butt) radius.

**Comment by yaldram** [comment]
Hello @jsartisan. Thanks for your review, yeah I knew but I thought this widget should match the Progress Bar one so copied the features over as it is. Sure I will remove all these as of now and keep it clean and simple. I will push the changes in a couple of hours.

**Comment by yaldram** [comment]
@jsartisan  I have removed the steps functionality and pushed the changes. Please check, thanks.

**Comment by jsartisan** [comment]
Checking in a bit.

**Comment by jsartisan** [comment]
Looks good to me.

**Comment by yaldram** [comment]
@jsartisan Hello, please can you review the final code. Thanks

**Comment by jsartisan** [comment]
@yaldram The PR looks ok to merge. We will merging this as soon as our release is unfrozen.

**Comment by jsartisan** [comment]
@yaldram can you check this review comment? 
image

**Comment by yaldram** [comment]
@jsartisan Thanks, I have added the property and pushed the changes, please check.

**Review by jsartisan — APPROVED** [review]

File: app/client/src/widgets/CircularProgressWidget/widget/index.tsx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      {
+        sectionName: "General",
+        children: [
+          {
+            propertyName: "progress",
+            helpText: "Sets the progress value of the widget",
+            label: "Progress",
+            controlType: "INPUT_TEXT",
+            placeholderText: "Value:",
+            isBindProperty: true,
```
[the remark this answers is not in the document]
Comment (yaldram) [reply app/client/src/widgets/CircularProgressWidget/widget/index.tsx:31]: I have added `isJSConvertible` property see line 33.

File: app/client/src/widgets/CircularProgressWidget/component/index.tsx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+  function drawPath() {
+    const rotation = counterClockwise ? 1 : 0;
+
+    // Move to center of canvas
+    // Relative move to top canvas
+    // Relative arc to bottom of canvas
+    // Relative arc to top of canvas
+    return `
+        M ${VIEWBOX_CENTER_X},${VIEWBOX_CENTER_Y}
```
[the remark this answers is not in the document]
Comment (yaldram) [reply app/client/src/widgets/CircularProgressWidget/component/index.tsx:68]: This is used to draw the trail and path, I took it from here - https://github.com/kevinsqi/react-circular-progressbar/blob/master/src/Path.tsx. It is drawing the path.

**Review by jsartisan — APPROVED** [review]

**Review by jsartisan — APPROVED** [review]

**Comment by jsartisan** [comment]
fix eslint typo

**Comment by jsartisan** [comment]
@yaldram One last thing, the build is failing for some reason. I believe it's because of eslint issue. Can you check?

**Comment by yaldram** [comment]
@jsartisan  Hey the build ran successfully on my local wsl setup.

**Comment by jsartisan** [comment]
@yaldram Can you come discord once? Let's have a call to fix the build issue.

**Comment by yaldram** [comment]
@jsartisan Ok, I am online on Discord.

**Review by jsartisan — APPROVED** [review]

**Comment by somangshu** [comment]
@allcontributors add @yaldram as a contributor for code, ideas

**Pull request merged**